### ISSUE TITLE
⛔ Elementor 4.2.0: Promotions module fatal error — Class "Ally_Dashboard_Widget" not found (site-wide 500)

### ISSUE THREAD

**Issue by costadelsol06** [elementor/elementor#36611]
### Description

Updating Elementor from 4.1.4 to **4.2.0** via `wp plugin update elementor` (WP-CLI) immediately breaks the entire site with a fatal PHP error, thrown on **every single request** — front-end, wp-admin, and admin-ajax alike. This is not limited to Elementor-edited pages; the whole site returns HTTP 500 the moment the plugin's `init` hook runs.

Fatal error (identical stack trace on front-end and wp-admin, only the entry file at the bottom of the trace differs):

```
PHP Fatal error:  Uncaught Error: Class "Elementor\Modules\Promotions\Widgets\Ally_Dashboard_Widget" not found in ../elementor/modules/promotions/module.php:101
Stack trace:
#0 ../elementor/core/base/module.php(86): Elementor\Modules\Promotions\Module->__construct()
#1 ../elementor/core/modules-manager.php(64): Elementor\Core\Base\Module::instance()
#2 ../elementor/includes/plugin.php(712): Elementor\Core\Modules_Manager->__construct()
#3 ../elementor/includes/plugin.php(630): Elementor\Plugin->init_components()
#4 ../wp-includes/class-wp-hook.php(341): Elementor\Plugin->init('')
#5 ../wp-includes/class-wp-hook.php(365): WP_Hook->apply_filters(NULL, Array)
#6 ../wp-includes/plugin.php(522): WP_Hook->do_action(Array)
#7 ../wp-settings.php(771): do_action('init')
#8 ../wp-config.php(140): require_once('../w...')
#9 ../wp-load.php(50): require_once('../w...')
#10 ../wp-blog-header.php(13): require_once('../w...')
#11 ../index.php(17): require('../w...')
#12 {main}
  thrown in ../elementor/modules/promotions/module.php on line 101
```

Site was down (500 on every URL, confirmed over both IPv4 and IPv6) from the moment the plugin finished updating. Recovered by immediately force-reinstalling the previous version: `wp plugin install elementor --version=4.1.4 --force`, which restored the site within ~2 minutes.

**This looks like a regression from [PR #36555](https://github.com/elementor/elementor/pull/36555) ("Internal: Remove Ally widget from WordPress dashboard page"), merged 2026-07-19 — one day before the 4.2.0 release.** That PR removed the `Ally_Dashboard_Widget` import/initialization and the `get_ally_external_scanner_url()` method from `modules/promotions/module.php`, describing it as "dead code removal with no dependencies... verified by isolated diff scope." Based on the fatal error above, a reference to `Ally_Dashboard_Widget` appears to remain somewhere the removal missed, so the class is still being instantiated/referenced even though its file/class definition is gone.

This also strongly resembles a previously-confirmed and fixed bug in this exact module: [#34659](https://github.com/elementor/elementor/issues/34659) ("Ally's Dashboard Widget causes 500 fatal error with undefined function"), which affected 3.35.2 and was fixed within hours in 3.35.3 — same widget, same module, same "fatal on every request" symptom, different specific undefined-symbol error.

Note: we independently also hit a very similar site-wide fatal (different missing class, `Birthday_Easter_Egg_Promotion`, also in `modules/promotions/module.php`) when 4.1.5 was released — which is why this site was pinned at 4.1.4 in the first place before this 4.2.0 test. Given the repeated history of this module (also see reverted PRs #34544/#34545, #34665/#34666 for the same Ally widget earlier this year), the Promotions module's init path seems to be a recurring source of production-breaking regressions.

### Steps to reproduce

1. Have a WordPress 7.0.2 site running Elementor 4.1.4 (free version, no Elementor Pro installed).
2. Update Elementor to 4.2.0 via WP-CLI: `wp plugin update elementor` (or via the standard wp-admin "Update Now" — same code path).
3. Load any front-end page, or `/wp-admin/`, or trigger `admin-ajax.php`.

### Expected behavior

The site should continue to function normally after updating Elementor to 4.2.0, the same as any routine plugin update. Instead, every single request fatals immediately because `Elementor\Modules\Promotions\Module`'s constructor unconditionally references `Ally_Dashboard_Widget`, a class that no longer exists after PR #36555's removal.

### Elementor System Info

<details>
<summary>Click to reveal</summary>

```txt


== Server Environment ==
	Operating System: Linux
	Software: Apache
	MySQL version: MariaDB Server v12.3.2-MariaDB
	PHP Version: 8.5.8
	PHP Memory Limit: -1
	PHP Max Input Vars: 10000
	PHP Max Post Size: 64M
	GD Installed: Yes
	ZIP Installed: Yes
	Write Permissions: All right
	Elementor Library: Connected

== WordPress Environment ==
	Version: 7.0.2
	WP Multisite: No
	Max Upload Size: 64 Mo
	Memory limit: 768M
	Max Memory limit: 768M
	Permalink Structure: /%year%/%monthnum%/%day%/%postname%/
	Language: fr_FR
	Timezone: Europe/Paris
	Debug Mode: Inactive

== Theme ==
	Name: Healing Vibes
	Version: 1.2.3
	Author: cmsmasters
	Child Theme: No

== User ==
	Role: administrator

== Active Plugins ==
	Advanced Custom Fields PRO 6.8.6
	Angie 1.1.11
	CMSMasters Elementor Addon 1.25.1
	CMSMasters Framework 1.0.20
	Custom Post Type UI 1.19.3
	Elementor 4.1.4  <-- crashed at 4.2.0 (subject of this report); force-reinstalled 4.1.4 to restore production, so this reads 4.1.4 again
	Envato Market 2.0.14
	Forminator 1.55.1
	Redis Object Cache 2.8.0
	Wordfence Security 8.2.2
	WP Mail SMTP 4.9.0

== Must-Use Plugins ==
	Blog Image Size (woocommerce_single) 1.0.0
	Disable XML-RPC (hardening) 1.0.0

== Elementor Experiments ==
	e_font_icon_svg: default (default: active)
	additional_custom_breakpoints: default (default: active)
	container: active (default: active)
	e_optimized_markup: default (default: active)
	e_panel_promotions: default (default: active)
	e_wp_abilities_api: default (default: inactive)
	e_pro_free_trial_popup: default (default: active)
	nested-elements: default (default: active)
	pages_panel: default (default: inactive)
	e_atomic_elements: default (default: active)
	e_indications_popover: default (default: inactive)
	atomic_widgets_should_enforce_capabilities: default (default: active)
	editor_mcp: default (default: active)
	e_bc_migrations: default (default: active)
	e_css_grid: default (default: inactive)
	e_editor_design_system_panel: default (default: active)
	e_classes: default (default: active)
	global_classes_should_enforce_capabilities: default (default: active)
	e_variables: default (default: active)
	e_variables_manager: default (default: active)
	e_opt_in_v4_page: default (default: active)
	e_opt_in_v4: default (default: active)
	e_components: default (default: active)
	e_interactions: default (default: active)
	e_widget_creation: default (default: active)
	markdown_rendering: default (default: inactive)
	import-export-customization: default (default: active)
	site-builder: default (default: inactive)
	in_editor_feedback: default (default: inactive)
	editor_events: default (default: active)



```

</details>

### Agreement

- [x] I confirm I have read and followed all the guidelines and instructions outlined in the Elementor Bug Report form.
- [x] I agree that my issue may be closed without further action if it doesn't meet all the requirements outlined in the Elementor Bug Report form.

**Comment by nicholaszein** [comment]
Hello @costadelsol06!

Thank you for submitting this issue! 🙏 

We have identified the problem and our team is already working on a fix.

Kind regards

**Comment by nicholaszein** [comment]
Hello @costadelsol06.

After further investigation we concluded this is not an issue caused by Elementor. It seems you have either OPCache or Object Cache storing the affected files in your server's system memory. 

In both PRs removing the Ally widget and the birthday easter egg, all references were correctly removed and are not longer part of our codebase. Which indicates your setup is likely caching `modules/promotions/module.php`.

Make sure you always purge and clear all levels of cache, including but not limited to OPCache and Object Cache.

For this reason we are closing this issue.

if you think this is still an issue caused by Elementor, and can confirm these files are not being cached by your system, pease follow the guidelines for opening a new [🐞 Bug Report](https://github.com/elementor/elementor/issues/new?assignees=&labels=status%2Fawaiting_triage&template=bug.yml), and include more information about your system environment that indicates the files are not being cached.

Kind regards
